lib/src/builder/build_step.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+library build.src.builder.build_step;
+
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+
+import '../asset/asset.dart';
+import '../asset/id.dart';
+import '../asset/reader.dart';
+import '../asset/writer.dart';
+
+/// A single step in the build processes. This represents a single input and
+/// its expected and real outputs. It also handles tracking of dependencies.
+class BuildStep {
+  /// The primary input for this build step.
+  final Asset input;
+
+  /// The list of all outputs which are expected/allowed to be output from this
+  /// step.
+  final List<AssetId> expectedOutputs;
+
+  /// The actual outputs of this build step.
+  UnmodifiableListView<Asset> get outputs => new UnmodifiableListView(_outputs);
```

Comment:
Definitely agree with both pieces of input here, I was already kind of thinking the same thing with the TODO on the next field and I think the same logic applies for this one. If only we had friend libraries I could easily leave these as private :).

Being that we don't have friend libraries, and would like to avoid parts, what do you think the best solution is? Create an InternalBuildStep class which wraps a BuildStep and exposes these private fields as public ones (but then hide it when exporting this library in `lib/build.dart`)? I would really rather not go the `part` route but that would solve this issue as well....